### PR TITLE
Fix hanging #37

### LINKED ISSUE #37: Hanging Line on Docker Build on Windows and OSX
The hanging line is: 

`RUN /bin/bash -c "cd e-mission-server && source setup/activate.sh && conda env update --name emission --file setup/environment36.notebook.additions.yml"`

Reproduced using docker containers and Anaconda environment on my home computer.

### PR DESCRIPTION
**Pull request by aGuttman**
Intended to fix #37 while using updated cloud approved server image. Messy, quick version of the solution. See issue comment for better solution outline.

### CODE REVIEW

File: viz_scripts/docker/Dockerfile [hunk 1]
```diff
@@ -1,8 +1,11 @@
 # python 3
-FROM emission/e-mission-server:2.9.2
+FROM emission/e-mission-server:4.0.0
 
 ADD environment36.dashboard.additions.yml /
 
+RUN /bin/bash -c "cd e-mission-server && cat setup/export_versions.sh && printf \"export EXP_CONDA_VER=4.12.0\" > setup/export_versions.sh && cat setup/export_versions.sh"
+RUN /bin/bash -c "cd e-mission-server && cat setup/environment36.notebook.additions.yml && printf \"name: emission\nchannels:\n- conda-forge\n- defaults\ndependencies:\n- branca=0.5.0\n- folium=0.12.1.post1\n- ipython=7.33.0\n- jedi=0.18.1\n- jupyter=1.0.0\n- matplotlib=3.5.1\n\" > setup/environment36.notebook.additions.yml && cat setup/environment36.notebook.additions.yml" 
+RUN /bin/bash -c "cd e-mission-server && source setup/setup_conda.sh Linux-x86_64 && source setup/setup.sh"
```
Comment (shankari) [inline viz_scripts/docker/Dockerfile:8]:
```suggestion
```

File: viz_scripts/docker/Dockerfile [hunk 2]
```diff
@@ -1,8 +1,11 @@
 # python 3
-FROM emission/e-mission-server:2.9.2
+FROM emission/e-mission-server:4.0.0
```
Comment (shankari) [inline viz_scripts/docker/Dockerfile:2]:
```suggestion
FROM emission/e-mission-server.dev.server-only:4.0.0
```

**Review by shankari — CHANGES_REQUESTED** [review]
Also add:

```
      - SERVER_REPO=https://github.com/aGuttman/e-mission-server.git
      - SERVER_BRANCH=dashboard-dependencies
```

to the `notebook-server`'s `environment` section so you can rebuild with your server changes.

Please test with these changes and confirm that it works before I merge everything

File: viz_scripts/docker/Dockerfile (hunk 2 above)
Comment (shankari) [reply viz_scripts/docker/Dockerfile:2]: Revert this after testing.

**Comment by aGuttman** [comment]
Tested with `emission/e-mission-server.dev.server-only:4.0.0` using updated dependancies from `github.com/aGuttman/e-mission-server.git#dashboard-dependencies`.

Works.

I'm unclear about what changes I should push. Asked to revert the `emission/e-mission-server.dev.server-only:4.0.0` change after testing and remove to in-place dependency updates from the docker file. Should I assume that `emission/e-mission-server:4.0.0` will have the updated dependancies and a working conda install?

**Comment by shankari** [comment]
@aGuttman can you write out the exact steps that you used for testing? This is complex, and I want to make sure that we actually did what we wanted to.

wrt

> Should I assume that emission/e-mission-server:4.0.0 will have the updated dependancies and a working conda install?

As you know from https://github.com/e-mission/e-mission-server/pull/875, `emission/e-mission-server:4.0.0` is a currently published repo which will not have the updated dependencies.

You need to make the changes that I requested in https://github.com/e-mission/e-mission-server/pull/875, we need to build a new image with the dependencies, and then bump this up.

And we will need to bump up the dependencies here periodically to upgrade to the most recent server image version, which we will presumably do as part of the CI for this repo.

**Comment by aGuttman** [comment]
In `Dockerfile`:
Replaced `FROM emission/e-mission-server:4.0.0` with `FROM emission/e-mission-server.dev.server-only:4.0.0`
Added clone `github.com/aGuttman/e-mission-server.git` and check out `dashboard-dependencies` (through env vars set in the docker-compose)
Removed in-place dependacy updates
```
RUN /bin/bash -c "cd e-mission-server && cat setup/export_versions.sh && printf \"export EXP_CONDA_VER=4.12.0\" > setup/export_versions.sh && cat setup/export_versions.sh"
RUN /bin/bash -c "cd e-mission-server && cat setup/environment36.notebook.additions.yml && printf \"name: emission\nchannels:\n- conda-forge\n- defaults\ndependencies:\n- branca=0.5.0\n- folium=0.12.1.post1\n- ipython=7.33.0\n- jedi=0.18.1\n- jupyter=1.0.0\n- matplotlib=3.5.1\n\" > setup/environment36.notebook.additions.yml && cat setup/environment36.notebook.additions.yml" 
```

Ran `docker-compose -f docker-compose.dev.yml  up`:
```
RUN /bin/bash -c "cd e-mission-server && source setup/setup_conda.sh Linux-x86_64 && source setup/setup.sh"
RUN /bin/bash -c "cd e-mission-server && source setup/activate.sh && conda env update --name emission --file setup/environment36.notebook.additions.yml"
RUN /bin/bash -c "cd e-mission-server && source setup/activate.sh && conda env update --name emission --file /environment36.dashboard.additions.yml"
```
without an issue. (Requested to remove `RUN /bin/bash -c "cd e-mission-server && source setup/setup_conda.sh Linux-x86_64 && source setup/setup.sh"` above, but conda is not on `emission/e-mission-server.dev.server-only:4.0.0`, not sure I can do anything without running this line?)
Container starts up without errors.

**Comment by shankari** [comment]
> Ran docker-compose -f docker-compose.dev.yml up:

you should also rebuild the `docker-compose` image. Otherwise, it will use the old image, right?
https://stackoverflow.com/a/39988980/4040267

> The following builds the images if the images do not exist and starts the containers:

**Comment by aGuttman** [comment]
> > Ran docker-compose -f docker-compose.dev.yml up:
> 
> you should also rebuild the `docker-compose` image. Otherwise, it will use the old image, right? https://stackoverflow.com/a/39988980/4040267
> 
> > The following builds the images if the images do not exist and starts the containers:

Correct, but I've been manually deleting the images. Forgot to mention that.

**Comment by shankari** [comment]
@aGuttman I can't actually merge this until you remove the hack
https://github.com/e-mission/em-public-dashboard/pull/53#discussion_r961203227

Also, do you have an ETA for addressing https://github.com/e-mission/e-mission-server/pull/875?
If it is not today, please plan on committing the changes you made for the testing (so that this is not broken) and reverting them when you enable CI for this repo.

**Comment by aGuttman** [comment]
I will have addressed e-mission/e-mission-server/pull/875 today, yes, before 2:30 hopefully.

**Comment by shankari** [comment]
Subsumed by https://github.com/e-mission/em-public-dashboard/pull/54

**Pull request closed without merging**